<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Plotting Tree with Data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<!-- Font Awesome -->
<!-- <link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" /> -->
<!-- <script src="https://use.fontawesome.com/54ee8c2dfd.js"></script> -->

<!-- Google fonts -->
<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,400italic,700,700italic|Oswald:400,700' rel='stylesheet' type='text/css'>

<!-- Favicon -->
<link rel="shortcut icon" type="image/x-icon" href="img/favicon.ico">

<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-77901140-1', 'auto');
  ga('send', 'pageview');
</script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="https:\\maxcdn.bootstrapcdn.com\font-awesome\4.5.0\css\font-awesome.min.css" type="text/css" />
<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">treedata-2019</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="setup.html">
    <span class="fa fa-cog"></span>
     
    Setup
  </a>
</li>
<li>
  <a href="data.html">
    <span class="fa fa-download"></span>
     
    Data
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Lessons
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">--- R &amp; ggplot2 ---</li>
    <li>
      <a href="r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="r-dataframes.html">Data Frames</a>
    </li>
    <li>
      <a href="r-viz-gapminder.html">Data Visualization</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">--- treeio ---</li>
    <li>
      <a href="r-treeio.html">Importing &amp; Exporting Trees with Data</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">--- ggtree ---</li>
    <li>
      <a href="r-ggtree-viz.html">Phylogenetic Tree Visualization</a>
    </li>
    <li>
      <a href="r-ggtree-annotation.html">Phylogenetic Tree Annotation</a>
    </li>
    <li>
      <a href="r-ggtree-data.html">Plotting Tree with Data</a>
    </li>
    <li>
      <a href="r-ggtree-phylomoji.html">Bonus! -  Phylomoji with ggtree</a>
    </li>
    <li>
      <a href="r-ggtree-image.html">Bonus! -  Phylogenetic tree with images</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="index.html#faq">FAQ</a>
    </li>
    <li>
      <a href="help.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/GuangchuangYu/treedata-workshop">Source code for this site</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Plotting Tree with Data</h1>

</div>


<p>This lesson demonstrates how to map and visualize associated data on phylogenetic tree using <strong>ggtree</strong>. The materials were modified from the <a href="https://academic.oup.com/mbe/article-abstract/35/12/3041/5142656">ggtree paper</a>.</p>
<p>This lesson covers two general methods for mapping and visualizing associated data on phylogeny. Method 1 allows external data to be mapped on the tree structure and used as visual characteristic in tree and data visualization. Method 2 plots the data with the tree side by side using different geometric functions after reordering the data based o the tree structure.</p>
<div id="mapping-data-to-the-tree-structure" class="section level2">
<h2>Mapping data to the tree structure</h2>
<p>One of the demands of mapping data is to link the data, such as phenotypic, experimental data and clinical data, to the tree structure, to display the data on the tree directly or use the data as visual characteristics of the tree branches/nodes. <strong>ggtree</strong> provides an operator, <code>%&lt;+%</code>, for attaching external data to the <code>ggtree</code> object. Any data frame that contains a column of “node” or first column of taxa labels can be integrated using the <code>%&lt;+%</code> operator. Multiple data sets can be attached progressively. When the data are attached, all the information stored in the data serve as numerical/categorical node attributes and can be directly used to visualize the tree by scaling the attributes as different colors or line sizes, label the tree using the original values of the attributes or parsing them as math expression, <a href="r-ggtree-phylomoji.html">emoji</a> or <a href="r-ggtree-image.html">silhouette image</a>.</p>
<p>This example uses the <code>%&lt;+%</code> operator to integrate taxon information and maps the data to different color and shape of symbolic points.</p>
<pre class="r"><code>library(ggplot2)
library(ggtree)

x &lt;- read.tree(&quot;data/tree_boots.nwk&quot;)
info &lt;- read.csv(&quot;data/tip_data.csv&quot;)
p &lt;- ggtree(x) %&lt;+% info + xlim(-.1, 5)
p2 &lt;- p + geom_tiplab(offset = 0.2) +
    geom_tippoint(aes(shape = trophic_habit, color = trophic_habit, size = mass_in_kg)) + 
    theme(legend.position = &quot;right&quot;) + scale_size_continuous(range = c(3, 10))
print(p2)</code></pre>
<p><img src="r-ggtree-data_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<div id="exercise-1" class="section level3">
<h3>Exercise 1</h3>
<ol style="list-style-type: decimal">
<li>Further integrate internal node data, and use posterior values to fill the background color to label internal nodes with vernacular names.</li>
</ol>
<p><img src="r-ggtree-data_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
</div>
</div>
<div id="align-graph-to-the-tree-based-on-tree-structure" class="section level2">
<h2>Align graph to the tree based on tree structure</h2>
<p>Evolutionary data are heterogeneous and many data types cannot be displayed on the tree directly, such as genetic information at a pan-genome scale, multiple sequence alignment, and species abundance distributions. The issue is that we need to reorder the plot based on the tree structure if we want to align the plot with the tree side by side to interpret the data in evolutionary context. This is quite challenging as tree structure is not human friendly and need expertise in programming. Consequenctly, <strong>ggtree</strong> provides the <code>facet_plot</code> function internally reorders the input data based on the tree structure and visualizes the data at the specific panel by the <code>geom</code> function.</p>
<p>The following example re-create <a href="img/F1.large.jpe">Figure 1</a> published on <a href="https://jvi.asm.org/content/91/21/e01285-17" class="uri">https://jvi.asm.org/content/91/21/e01285-17</a>.</p>
<pre class="r"><code>library(Biostrings)
library(tidyr)
library(ggplot2)
library(ggtree)

tree &lt;- read.tree(&quot;data/HPV58.tree&quot;)

clade &lt;- c(A3 = 92, A1 = 94, A2 = 108, B1 = 156, B2 = 159, C = 163, D1 = 173, D2 = 176)
tree &lt;- groupClade(tree, clade)

## visualize the tree with tip labels and tree scale
p &lt;- ggtree(tree, aes(color = group), ladderize = FALSE) 

tipseq_aln &lt;- readDNAStringSet(&quot;data/HPV58_aln.fas&quot;)


## calculate pairwise hamming distances among sequences
tipseq_dist &lt;- stringDist(tipseq_aln, method = &quot;hamming&quot;)

## calculate percentage of differences
tipseq_d &lt;- as.matrix(tipseq_dist) / width(tipseq_aln[1]) * 100

## convert the matrix to tidy data frame for facet_plot
dd &lt;- as.data.frame(tipseq_d)
dd$seq1 &lt;- rownames(tipseq_d)
td &lt;- gather(dd,seq2, dist, -seq1)

## visualize the sequence differences using dot plot and line plot
## and align the sequence difference plot to the tree using facet_plot
p2 &lt;- facet_plot(p, panel = &quot;Sequence Distance&quot;, data = td, geom_point, 
            mapping = aes(x = dist, color = group), alpha = .6) %&gt;%
    facet_plot(panel = &quot;Sequence Distance&quot;, data = td, geom = geom_path, 
            mapping=aes(x = dist, group = seq2, color = group), alpha = .6)

print(p2)</code></pre>
<p><img src="r-ggtree-data_files/figure-html/jv-1.png" width="864" /></p>
<div id="exercise-2" class="section level3">
<h3>Exercise 2</h3>
<ol style="list-style-type: decimal">
<li>create a dummy data and align a barplot (need <code>ggstance::geom_barh</code> layer) with the <code>p2</code> show above.</li>
</ol>
<p><img src="r-ggtree-data_files/figure-html/unnamed-chunk-4-1.png" width="1152" /></p>
</div>
</div>

<div class="footer">
<a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank"><i class="fa fa-creative-commons fa-2x"></i></a>&nbsp;
<a href="https://twitter.com/guangchuangyu" target="_blank"><i class="fa fa-twitter fa-2x"></i></a>&nbsp;
<a href="https://github.com/GuangchuangYu/treedata-workshop" target="_blank"><i class="fa fa-github fa-2x"></i></a>&nbsp;
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
