<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Importing &amp; Exporting Trees with Data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<!-- Font Awesome -->
<!-- <link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" /> -->
<!-- <script src="https://use.fontawesome.com/54ee8c2dfd.js"></script> -->

<!-- Google fonts -->
<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,400italic,700,700italic|Oswald:400,700' rel='stylesheet' type='text/css'>

<!-- Favicon -->
<link rel="shortcut icon" type="image/x-icon" href="img/favicon.ico">

<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-77901140-1', 'auto');
  ga('send', 'pageview');
</script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="https:\\maxcdn.bootstrapcdn.com\font-awesome\4.5.0\css\font-awesome.min.css" type="text/css" />
<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">treedata-2019</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="setup.html">
    <span class="fa fa-cog"></span>
     
    Setup
  </a>
</li>
<li>
  <a href="data.html">
    <span class="fa fa-download"></span>
     
    Data
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Lessons
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">--- R &amp; ggplot2 ---</li>
    <li>
      <a href="r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="r-dataframes.html">Data Frames</a>
    </li>
    <li>
      <a href="r-viz-gapminder.html">Data Visualization</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">--- treeio ---</li>
    <li>
      <a href="r-treeio.html">Importing &amp; Exporting Trees with Data</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">--- ggtree ---</li>
    <li>
      <a href="r-ggtree-viz.html">Phylogenetic Tree Visualization</a>
    </li>
    <li>
      <a href="r-ggtree-annotation.html">Phylogenetic Tree Annotation</a>
    </li>
    <li>
      <a href="r-ggtree-phylomoji.html">Bonus! -  Phylomoji with ggtree</a>
    </li>
    <li>
      <a href="r-ggtree-images.html">Bonus! -  Phylogenetic tree with images</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="index.html#faq">FAQ</a>
    </li>
    <li>
      <a href="help.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/GuangchuangYu/treedata-workshop">Source code for this site</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Importing &amp; Exporting Trees with Data</h1>

</div>


<p>This section will demonstrate how to use <strong>treeio</strong> to parse tree with associated data to a single object in R.</p>
<p><em>This lesson assumes a <a href="r-basics.html">basic familiarity with R</a> and <a href="r-dataframes.html">data frames</a>.</em></p>
<p>This lesson does <em>not</em> cover methods and software for <em>generating</em> phylogenetic trees, nor does it it cover <em>interpreting</em> phylogenies. <strong><a href="https://link.springer.com/article/10.1007/s12052-010-0273-6">Here’s a quick primer on how to read a phylogeny</a></strong> that you should definitely review prior to this lesson, but it is by no means extensive. Genome-wide sequencing allows for examination of the entire genome, and from this, many methods and software tools exist for comparative genomics using SNP- and gene-based phylogenetic analysis, either from unassembled sequencing reads, draft assemblies/contigs, or complete genome sequences. These methods are beyond the scope of this lesson.</p>
<div id="the-treeio-package" class="section level2">
<h2>The <code>treeio</code> Package</h2>
<p><strong>treeio</strong> is an R package that designed for phylogenetic tree data input and output. It is released as part of Bioconductor and ROpenSci projects.</p>
<ol style="list-style-type: decimal">
<li><strong>tree Bioconductor page</strong>: <a href="https://www.bioconductor.org/packages/treeio" class="uri">https://www.bioconductor.org/packages/treeio</a>.</li>
<li><strong>treeio homepage</strong>: <a href="https://guangchuangyu.github.io/treeio" class="uri">https://guangchuangyu.github.io/treeio</a>.</li>
</ol>
<p>Just like R packages from CRAN, you only need to install Bioconductor packages once (<a href="setup.html#bioconductor">instructions here</a>), then load them every time you start a new R session.</p>
<pre class="r"><code>library(treeio)</code></pre>
<p>Most tree viewer software (including R packages) focus on <strong>Newick</strong> and <strong>NEXUS</strong> file formats, and other evolutionary analysis software might also contains supporting evidence and/or analysis findings within the file that can be further analyzed in R or interpreted in phylogenetic context to help identifying evolutionary patterns.</p>
<p><strong>treeio</strong> supports several file formats, including:</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Newick_format">Newick</a></li>
<li><a href="https://en.wikipedia.org/wiki/Nexus_file">Nexus</a></li>
<li><a href="https://home.cc.umanitoba.ca/~psgendb/doc/atv/NHX.pdf">New Hampshire eXtended format (NHX)</a></li>
<li><a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0031009">jplace</a></li>
<li><a href="https://en.wikipedia.org/wiki/PHYLIP">Phylip</a></li>
</ul>
<p>and software output from:</p>
<ul>
<li><a href="http://beast2.org/">BEAST</a></li>
<li><a href="http://sco.h-its.org/exelixis/web/software/epa/index.html">EPA</a></li>
<li><a href="https://veg.github.io/hyphy-site/">HyPhy</a></li>
<li><a href="http://nbisweden.github.io/MrBayes/">MrBayes</a></li>
<li><a href="http://abacus.gene.ucl.ac.uk/software/paml.html">PAML</a></li>
<li><a href="https://pbil.univ-lyon1.fr/software/phyldog/">PHYLDOG</a></li>
<li><a href="http://matsen.fhcrc.org/pplacer/">pplacer</a></li>
<li><a href="http://ginger.ucdavis.edu/r8s">r8s</a></li>
<li><a href="http://evomics.org/learning/phylogenetics/raxml/">RAxML</a></li>
<li><a href="https://revbayes.github.io/intro.html">RevBayes</a></li>
</ul>
<hr />
</div>
<div id="basic-trees" class="section level2">
<h2>Basic trees</h2>
<p>Let’s first import a phylogenetic tree in to R. Phylogenetic tree are mainly stored as Newick or Nexus formats that only contains tree structure with (phylogram) or without (cladogram) branch lengths.</p>
<p>Download the <a href="data/tree.nwk"><strong>tree.nwk</strong> data by clicking here</a> or using the link above. Let’s load the libraries you’ll need if you haven’t already, and then import the tree using <code>read.tree()</code>. Displaying the object itself tells you a little bit about the tree, e.g. number of tips and nodes, a glance of tip (and node if available) labels.</p>
<pre class="r"><code>library(treeio)
tree &lt;- read.tree(&quot;data/tree.nwk&quot;)
tree</code></pre>
<pre><code>## 
## Phylogenetic tree with 28 tips and 26 internal nodes.
## 
## Tip labels:
##  Phy000G05U_EMENI, Phy000GDP6_ASPNG, Phy003AMS0_602072, Phy000FJDH_ASPFL, Phy000FCLK_ASPCL, Phy000FQ5O_ASPFU, ...
## Node labels:
##  , 0.99985, 0.99985, 0.72129, 0.991353, 0.99985, ...
## 
## Unrooted; includes branch lengths.</code></pre>
<p>The only possible way for a Newick tree to store additional information other than the tree structure is to encode the information as taxa labels. This tree use node label to store support values (<em>e.g.</em> bootstrap values).</p>
<div id="exercise-1" class="section level3">
<h3>Exercise 1</h3>
<p>Look at the help page for <code>?ape::read.tree</code>, specifically the <strong>Details</strong> and <strong>Value</strong> session to explore the components of the <code>phylo</code> object.</p>
<ol style="list-style-type: decimal">
<li>Summarize a branch length distribution by Histogram.</li>
</ol>
<hr />
</div>
</div>
<div id="getting-tree-data-from-evolutionary-analysis-result" class="section level2">
<h2>Getting tree data from evolutionary analysis result</h2>
<p>The <code>treeio</code> package implement several parser functions.</p>
<table>
<caption>Parser functions defined in treeio</caption>
<thead>
<tr class="header">
<th align="left">Parser function</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">read.beast</td>
<td align="left">parsing output of BEAST</td>
</tr>
<tr class="even">
<td align="left">read.codeml</td>
<td align="left">parsing output of CodeML (rst and mlc files)</td>
</tr>
<tr class="odd">
<td align="left">read.codeml_mlc</td>
<td align="left">parsing mlc file (output of CodeML)</td>
</tr>
<tr class="even">
<td align="left">read.hyphy</td>
<td align="left">parsing output of HYPHY</td>
</tr>
<tr class="odd">
<td align="left">read.jplace</td>
<td align="left">parsing jplace file including output of EPA and pplacer</td>
</tr>
<tr class="even">
<td align="left">read.mrbayes</td>
<td align="left">parsing output of MrBayes</td>
</tr>
<tr class="odd">
<td align="left">read.newick</td>
<td align="left">parsing newick string, with ability to parse node label as support values</td>
</tr>
<tr class="even">
<td align="left">read.nhx</td>
<td align="left">parsing NHX file including output of PHYLDOG and RevBayes</td>
</tr>
<tr class="odd">
<td align="left">read.paml_rst</td>
<td align="left">parsing rst file (output of BaseML or CodeML)</td>
</tr>
<tr class="even">
<td align="left">read.phylip</td>
<td align="left">parsing phylip file (phylip alignment + newick string)</td>
</tr>
<tr class="odd">
<td align="left">read.r8s</td>
<td align="left">parsing output of r8s</td>
</tr>
<tr class="even">
<td align="left">read.raxml</td>
<td align="left">parsing output of RAxML</td>
</tr>
</tbody>
</table>
<p>After parsing, storage of the tree structure with associated data is made through a S4 class, treedata, defined in the <a href="https://bioconductor.org/packages/treeio/">treeio</a> package. These parsed data are mapped to the tree branches and nodes inside <code>treedata</code> object, so that they can be efficiently used to visually annotate the tree using <a href="https://bioconductor.org/packages/ggtree/">ggtree</a> package.</p>
<p>Here, we use BEAST output as an example of using these parser functions to import tree with associated data. The details of BEAST output can be found on <a href="http://beast.community/nexus_metacomments.html" class="uri">http://beast.community/nexus_metacomments.html</a>. In summary, it introudces ‘metacomment’ to annotate elements in the standard NEXUS format. The additional information is put in comments so that existing programs can read the tree by ignoring them. <strong>treeio</strong> is able to read the tree with all inserted information stored in BEAST output.</p>
<pre class="r"><code>beast_tree &lt;- read.beast(&quot;data/MCC_FluA_H3.tree&quot;)
beast_tree</code></pre>
<pre><code>## &#39;treedata&#39; S4 object that stored information of
##  &#39;data/MCC_FluA_H3.tree&#39;.
## 
## ...@ phylo: 
## Phylogenetic tree with 76 tips and 75 internal nodes.
## 
## Tip labels:
##  A/Hokkaido/30-1-a/2013, A/New_York/334/2004, A/New_York/463/2005, A/New_York/452/1999, A/New_York/238/2005, A/New_York/523/1998, ...
## 
## Rooted; includes branch lengths.
## 
## with the following features available:
##  &#39;height&#39;,   &#39;height_0.95_HPD&#39;,  &#39;height_median&#39;,    &#39;height_range&#39;, &#39;length&#39;,
##  &#39;length_0.95_HPD&#39;,  &#39;length_median&#39;,    &#39;length_range&#39;, &#39;posterior&#39;,    &#39;rate&#39;,
##  &#39;rate_0.95_HPD&#39;,    &#39;rate_median&#39;,  &#39;rate_range&#39;.</code></pre>
<p>As there are many R package works on the <code>phylo</code> object, <code>treeio</code> provides <code>as.phylo</code> method to convert <code>treedata</code> object to a <code>phylo</code> object, and <code>get.data</code> method to extract associated data stored in <code>treedata</code> object.</p>
<p>The <a href="https://cran.r-project.org/web/packages/tidytree/index.html">tidytree</a> provides <code>as_tibble</code> method to convert <code>treedata</code> object to tidy data and creates the possibility of manipulating tree data using tidy interface.</p>
<div id="exercise-2" class="section level3">
<h3>Exercise 2</h3>
<ol style="list-style-type: decimal">
<li>convert the <code>beast_tree</code> to <code>phylo</code> object.</li>
<li>extract tree associated data from <code>beast_tree</code>.</li>
<li>convert <code>beast_tree</code> to tidy data.</li>
<li>create a scatter plot of substitution rate vs branch length of the <code>beast_tree</code></li>
</ol>
</div>
</div>
<div id="linking-external-data-to-phylogeny" class="section level2">
<h2>Linking external data to phylogeny</h2>
<p>There is a wide range of heterogeneous data, such as traits, geographic distritution, experimental and clinical data, <em>etc</em>, that need to be integrated and linked to phylogeny. For example, in the study of viral evolution, tree nodes may associated with epidemiological information, such as location, age and subtype. Functional annotations may need to be mapped on gene trees for comparative genomic studies.</p>
<p>To facilitate data integration, <strong>treeio</strong> provides <code>full_join</code> method to link external data to phylogeny and stored in <code>treedata</code> object.</p>
<pre class="r"><code>library(readr)

x &lt;- read.tree(&quot;data/tree_boots.nwk&quot;)

info &lt;- read_csv(&quot;data/tip_data.csv&quot;)
names(info)[1] &lt;- &#39;label&#39;
print(info)</code></pre>
<pre><code>## # A tibble: 7 x 10
##   label vernacularName imageURL imageLicense imageAuthor infoURL mass_in_kg
##   &lt;chr&gt; &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;        &lt;chr&gt;       &lt;chr&gt;        &lt;dbl&gt;
## 1 Rang~ Reindeer       http://~ CC-BY-SA     Alexandre ~ http:/~     109.  
## 2 Cerv~ Red deer       http://~ CC-BY-SA     Sciadopitys http:/~     241.  
## 3 Bos_~ Cattle         https:/~ CC-BY-SA     Cynthia Si~ http:/~     619.  
## 4 Ovis~ Asiatic moufl~ http://~ CC-BY-SA     J&lt;U+0161&gt;rg Hempel http:/~      39.1 
## 5 Suri~ Meerkat        http://~ CC-BY-SA     Sara&amp;Joach~ http:/~       0.73
## 6 Cyst~ Hooded seal    http://~ CC-BY-SA     Ecomare, S~ http:/~     279.  
## 7 Meph~ Striped skunk  http://~ CC-BY        Kevin Bowm~ http:/~       2.4 
## # ... with 3 more variables: trophic_habit &lt;chr&gt;, ncbi_taxid &lt;dbl&gt;,
## #   rank &lt;chr&gt;</code></pre>
<pre class="r"><code>d2 &lt;- read_csv(&quot;data/inode_data.csv&quot;)
names(d2)[1] &lt;- &#39;label&#39;

tree2 &lt;- full_join(as.treedata(x), info, by=&#39;label&#39;)
tree2</code></pre>
<pre><code>## &#39;treedata&#39; S4 object&#39;.
## 
## ...@ phylo: 
## Phylogenetic tree with 7 tips and 6 internal nodes.
## 
## Tip labels:
##  Rangifer_tarandus, Cervus_elaphus, Bos_taurus, Ovis_orientalis, Suricata_suricatta, Cystophora_cristata, ...
## Node labels:
## [1] &quot;Mammalia&quot;     &quot;Artiodactyla&quot; &quot;Cervidae&quot;     &quot;Bovidae&quot;     
## [5] &quot;Carnivora&quot;    &quot;Caniformia&quot;  
## 
## Rooted; includes branch lengths.
## 
## with the following features available:
##  &#39;&#39;, &#39;vernacularName&#39;,   &#39;imageURL&#39;, &#39;imageLicense&#39;, &#39;imageAuthor&#39;,
##  &#39;infoURL&#39;,  &#39;mass_in_kg&#39;,   &#39;trophic_habit&#39;,    &#39;ncbi_taxid&#39;,   &#39;rank&#39;.</code></pre>
<pre class="r"><code>full_join(tree2, d2, by=&#39;label&#39;)</code></pre>
<pre><code>## &#39;treedata&#39; S4 object&#39;.
## 
## ...@ phylo: 
## Phylogenetic tree with 7 tips and 6 internal nodes.
## 
## Tip labels:
##  Rangifer_tarandus, Cervus_elaphus, Bos_taurus, Ovis_orientalis, Suricata_suricatta, Cystophora_cristata, ...
## Node labels:
## [1] &quot;Mammalia&quot;     &quot;Artiodactyla&quot; &quot;Cervidae&quot;     &quot;Bovidae&quot;     
## [5] &quot;Carnivora&quot;    &quot;Caniformia&quot;  
## 
## Rooted; includes branch lengths.
## 
## with the following features available:
##  &#39;&#39;, &#39;vernacularName.x&#39;, &#39;imageURL&#39;, &#39;imageLicense&#39;, &#39;imageAuthor&#39;,
##  &#39;infoURL.x&#39;,    &#39;mass_in_kg&#39;,   &#39;trophic_habit&#39;,    &#39;ncbi_taxid&#39;,   &#39;rank.x&#39;,
##  &#39;vernacularName.y&#39;, &#39;infoURL.y&#39;,    &#39;rank.y&#39;,   &#39;bootstrap&#39;,    &#39;posterior&#39;.</code></pre>
<div id="exercise-3" class="section level3">
<h3>Exercise 3</h3>
<ol style="list-style-type: decimal">
<li>link internal node data (<code>d2</code>) to tree (<code>tree2</code>)</li>
<li>check whether you have mapped internal node data successful (<em>e.g.</em> bootstrap and posterior)</li>
<li>create a scatter plot of posterior vs bootstrap values.</li>
</ol>
<hr />
</div>
</div>
<div id="exporting-trees-with-data" class="section level2">
<h2>Exporting trees with data</h2>
<p><strong>treeio</strong> provides <code>write.beast</code> and <code>write.jtree</code> to export tree with data to a single file. <code>write.beast</code> outputs BEAST compatible file that can be open using <a href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> and many other software - most of them can read the tree by ignoring data.</p>
<pre class="r"><code>write.beast(tree2)</code></pre>
<pre><code>## #NEXUS
## [R-package treeio, Fri Jan 04 17:52:44 2019]
## 
## BEGIN TAXA;
##  DIMENSIONS NTAX = 7;
##  TAXLABELS
##      Rangifer_tarandus
##      Cervus_elaphus
##      Bos_taurus
##      Ovis_orientalis
##      Suricata_suricatta
##      Cystophora_cristata
##      Mephitis_mephitis
##  ;
## END;
## BEGIN TREES;
##  TRANSLATE
##      1   Rangifer_tarandus,
##      2   Cervus_elaphus,
##      3   Bos_taurus,
##      4   Ovis_orientalis,
##      5   Suricata_suricatta,
##      6   Cystophora_cristata,
##      7   Mephitis_mephitis
##  ;
##  TREE * UNTITLED = [&amp;R] (((1[&amp;vernacularName=Reindeer,imageURL=http://media.eol.org/content/2012/06/13/00/48543_orig.jpg,imageLicense=CC-BY-SA,imageAuthor=Alexandre Buisse (Nattfodd),infoURL=http://eol.org/pages/328653/overview,mass_in_kg=109.09,trophic_habit=herbivore,ncbi_taxid=9870,rank=species]:1,2[&amp;vernacularName=Red deer,imageURL=http://media.eol.org/content/2014/09/16/00/20239_orig.jpg,imageLicense=CC-BY-SA,imageAuthor=Sciadopitys,infoURL=http://eol.org/pages/328649/overview,mass_in_kg=240.87,trophic_habit=herbivore,ncbi_taxid=9860,rank=species]:1)Cervidae:1,(3[&amp;vernacularName=Cattle,imageURL=https://media.eol.org/content/2014/09/29/06/46535_orig.jpg,imageLicense=CC-BY-SA,imageAuthor=Cynthia Sims Parr,infoURL=http://eol.org/pages/328699/overview,mass_in_kg=618.64,trophic_habit=herbivore,ncbi_taxid=9913,rank=species]:1,4[&amp;vernacularName=Asiatic mouflon,imageURL=http://media.eol.org/content/2015/05/20/03/80720_orig.jpg,imageLicense=CC-BY-SA,imageAuthor=J&lt;U+0161&gt;rg Hempel,infoURL=http://eol.org/pages/311906/overview,mass_in_kg=39.1,trophic_habit=herbivore,ncbi_taxid=469796,rank=species]:1)Bovidae:1)Artiodactyla:1,(5[&amp;vernacularName=Meerkat,imageURL=http://media.eol.org/content/2016/08/16/05/67138_orig.jpg,imageLicense=CC-BY-SA,imageAuthor=Sara&amp;Joachim,infoURL=http://eol.org/pages/311580/overview,mass_in_kg=0.73,trophic_habit=carnivore,ncbi_taxid=37032,rank=species]:2,(6[&amp;vernacularName=Hooded seal,imageURL=http://media.eol.org/content/2013/06/18/07/63362_orig.jpg,imageLicense=CC-BY-SA,imageAuthor=Ecomare, Salko de Wolf,infoURL=http://eol.org/pages/328632/overview,mass_in_kg=278.9,trophic_habit=omnivore,ncbi_taxid=39293,rank=species]:1,7[&amp;vernacularName=Striped skunk,imageURL=http://media.eol.org/content/2012/06/15/06/75234_orig.jpg,imageLicense=CC-BY,imageAuthor=Kevin Bowman,infoURL=http://eol.org/pages/328593/overview,mass_in_kg=2.4,trophic_habit=omnivore,ncbi_taxid=30548,rank=species]:1)Caniformia:1)Carnivora:1)Mammalia;
## END;</code></pre>
<p><code>write.jtree</code> outputs JSON (JavaScript Object Notation) file, which is a lightweight data-interchange format and widely supported in almost all modern programming languages.</p>
<pre class="r"><code>write.jtree(tree2)</code></pre>
<pre><code>## {
##  &quot;tree&quot;: &quot;(((Rangifer_tarandus:1{1},Cervus_elaphus:1{2})Cervidae:1{10},(Bos_taurus:1{3},Ovis_orientalis:1{4})Bovidae:1{11})Artiodactyla:1{9},(Suricata_suricatta:2{5},(Cystophora_cristata:1{6},Mephitis_mephitis:1{7})Caniformia:1{13})Carnivora:1{12})Mammalia{8};&quot;,
##  &quot;data&quot;:[
##   {
##     &quot;edge_num&quot;: 1,
##     &quot;vernacularName&quot;: &quot;Reindeer&quot;,
##     &quot;imageURL&quot;: &quot;http://media.eol.org/content/2012/06/13/00/48543_orig.jpg&quot;,
##     &quot;imageLicense&quot;: &quot;CC-BY-SA&quot;,
##     &quot;imageAuthor&quot;: &quot;Alexandre Buisse (Nattfodd)&quot;,
##     &quot;infoURL&quot;: &quot;http://eol.org/pages/328653/overview&quot;,
##     &quot;mass_in_kg&quot;: 109.09,
##     &quot;trophic_habit&quot;: &quot;herbivore&quot;,
##     &quot;ncbi_taxid&quot;: 9870,
##     &quot;rank&quot;: &quot;species&quot;
##   },
##   {
##     &quot;edge_num&quot;: 2,
##     &quot;vernacularName&quot;: &quot;Red deer&quot;,
##     &quot;imageURL&quot;: &quot;http://media.eol.org/content/2014/09/16/00/20239_orig.jpg&quot;,
##     &quot;imageLicense&quot;: &quot;CC-BY-SA&quot;,
##     &quot;imageAuthor&quot;: &quot;Sciadopitys&quot;,
##     &quot;infoURL&quot;: &quot;http://eol.org/pages/328649/overview&quot;,
##     &quot;mass_in_kg&quot;: 240.87,
##     &quot;trophic_habit&quot;: &quot;herbivore&quot;,
##     &quot;ncbi_taxid&quot;: 9860,
##     &quot;rank&quot;: &quot;species&quot;
##   },
##   {
##     &quot;edge_num&quot;: 3,
##     &quot;vernacularName&quot;: &quot;Cattle&quot;,
##     &quot;imageURL&quot;: &quot;https://media.eol.org/content/2014/09/29/06/46535_orig.jpg&quot;,
##     &quot;imageLicense&quot;: &quot;CC-BY-SA&quot;,
##     &quot;imageAuthor&quot;: &quot;Cynthia Sims Parr&quot;,
##     &quot;infoURL&quot;: &quot;http://eol.org/pages/328699/overview&quot;,
##     &quot;mass_in_kg&quot;: 618.64,
##     &quot;trophic_habit&quot;: &quot;herbivore&quot;,
##     &quot;ncbi_taxid&quot;: 9913,
##     &quot;rank&quot;: &quot;species&quot;
##   },
##   {
##     &quot;edge_num&quot;: 4,
##     &quot;vernacularName&quot;: &quot;Asiatic mouflon&quot;,
##     &quot;imageURL&quot;: &quot;http://media.eol.org/content/2015/05/20/03/80720_orig.jpg&quot;,
##     &quot;imageLicense&quot;: &quot;CC-BY-SA&quot;,
##     &quot;imageAuthor&quot;: &quot;J&lt;U+0161&gt;rg Hempel&quot;,
##     &quot;infoURL&quot;: &quot;http://eol.org/pages/311906/overview&quot;,
##     &quot;mass_in_kg&quot;: 39.1,
##     &quot;trophic_habit&quot;: &quot;herbivore&quot;,
##     &quot;ncbi_taxid&quot;: 469796,
##     &quot;rank&quot;: &quot;species&quot;
##   },
##   {
##     &quot;edge_num&quot;: 5,
##     &quot;vernacularName&quot;: &quot;Meerkat&quot;,
##     &quot;imageURL&quot;: &quot;http://media.eol.org/content/2016/08/16/05/67138_orig.jpg&quot;,
##     &quot;imageLicense&quot;: &quot;CC-BY-SA&quot;,
##     &quot;imageAuthor&quot;: &quot;Sara&amp;Joachim&quot;,
##     &quot;infoURL&quot;: &quot;http://eol.org/pages/311580/overview&quot;,
##     &quot;mass_in_kg&quot;: 0.73,
##     &quot;trophic_habit&quot;: &quot;carnivore&quot;,
##     &quot;ncbi_taxid&quot;: 37032,
##     &quot;rank&quot;: &quot;species&quot;
##   },
##   {
##     &quot;edge_num&quot;: 6,
##     &quot;vernacularName&quot;: &quot;Hooded seal&quot;,
##     &quot;imageURL&quot;: &quot;http://media.eol.org/content/2013/06/18/07/63362_orig.jpg&quot;,
##     &quot;imageLicense&quot;: &quot;CC-BY-SA&quot;,
##     &quot;imageAuthor&quot;: &quot;Ecomare, Salko de Wolf&quot;,
##     &quot;infoURL&quot;: &quot;http://eol.org/pages/328632/overview&quot;,
##     &quot;mass_in_kg&quot;: 278.9,
##     &quot;trophic_habit&quot;: &quot;omnivore&quot;,
##     &quot;ncbi_taxid&quot;: 39293,
##     &quot;rank&quot;: &quot;species&quot;
##   },
##   {
##     &quot;edge_num&quot;: 7,
##     &quot;vernacularName&quot;: &quot;Striped skunk&quot;,
##     &quot;imageURL&quot;: &quot;http://media.eol.org/content/2012/06/15/06/75234_orig.jpg&quot;,
##     &quot;imageLicense&quot;: &quot;CC-BY&quot;,
##     &quot;imageAuthor&quot;: &quot;Kevin Bowman&quot;,
##     &quot;infoURL&quot;: &quot;http://eol.org/pages/328593/overview&quot;,
##     &quot;mass_in_kg&quot;: 2.4,
##     &quot;trophic_habit&quot;: &quot;omnivore&quot;,
##     &quot;ncbi_taxid&quot;: 30548,
##     &quot;rank&quot;: &quot;species&quot;
##   },
##   {
##     &quot;edge_num&quot;: 8
##   },
##   {
##     &quot;edge_num&quot;: 9
##   },
##   {
##     &quot;edge_num&quot;: 10
##   },
##   {
##     &quot;edge_num&quot;: 11
##   },
##   {
##     &quot;edge_num&quot;: 12
##   },
##   {
##     &quot;edge_num&quot;: 13
##   }
## ],
##  &quot;metadata&quot;: {&quot;info&quot;: &quot;R-package treeio&quot;, &quot;data&quot;: &quot;Fri Jan 04 17:52:44 2019&quot;}
## }</code></pre>
<div id="exercise-4" class="section level3">
<h3>Exercise 4</h3>
<ol style="list-style-type: decimal">
<li>output the file to a single file by passing <code>file = "your_file_name"</code> to <code>write.beast</code> or <code>write.jtree</code>.</li>
<li>re-read the output file into R using <code>read.beast</code> or <code>read.jtree</code>.</li>
</ol>
</div>
</div>

<div class="footer">
<a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank"><i class="fa fa-creative-commons fa-2x"></i></a>&nbsp;
<a href="https://twitter.com/guangchuangyu" target="_blank"><i class="fa fa-twitter fa-2x"></i></a>&nbsp;
<a href="https://github.com/GuangchuangYu/treedata-workshop" target="_blank"><i class="fa fa-github fa-2x"></i></a>&nbsp;
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
