<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Importing trees with data using treeio</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<!-- Font Awesome -->
<!-- <link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" /> -->
<!-- <script src="https://use.fontawesome.com/54ee8c2dfd.js"></script> -->

<!-- Google fonts -->
<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,400italic,700,700italic|Oswald:400,700' rel='stylesheet' type='text/css'>

<!-- Favicon -->
<link rel="shortcut icon" type="image/x-icon" href="img/favicon.ico">

<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-77901140-1', 'auto');
  ga('send', 'pageview');
</script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="https:\\maxcdn.bootstrapcdn.com\font-awesome\4.5.0\css\font-awesome.min.css" type="text/css" />
<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">treedata-2019</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="setup.html">
    <span class="fa fa-cog"></span>
     
    Setup
  </a>
</li>
<li>
  <a href="data.html">
    <span class="fa fa-download"></span>
     
    Data
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Lessons
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">--- R &amp; ggplot2 ---</li>
    <li>
      <a href="r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="r-dataframes.html">Data Frames</a>
    </li>
    <li>
      <a href="r-viz-gapminder.html">Data Visualization</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">--- treeio ---</li>
    <li>
      <a href="r-treeio-input.html">Importing trees with data using treeio</a>
    </li>
    <li>
      <a href="r-treeio-output.html">Exporting trees with data using treeio</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">--- ggtree ---</li>
    <li>
      <a href="r-ggtree.html">Phylogenetic Trees</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="index.html#faq">FAQ</a>
    </li>
    <li>
      <a href="help.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/GuangchuangYu/treedata-workshop">Source code for this site</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Importing trees with data using treeio</h1>

</div>


<p>This section will demonstrate how to use <strong>treeio</strong> to parse tree with associated data to a single object in R.</p>
<p><em>This lesson assumes a <a href="r-basics.html">basic familiarity with R</a> and <a href="r-dataframes.html">data frames</a>.</em></p>
<p>This lesson does <em>not</em> cover methods and software for <em>generating</em> phylogenetic trees, nor does it it cover <em>interpreting</em> phylogenies. <strong><a href="https://link.springer.com/article/10.1007/s12052-010-0273-6">Here’s a quick primer on how to read a phylogeny</a></strong> that you should definitely review prior to this lesson, but it is by no means extensive. Genome-wide sequencing allows for examination of the entire genome, and from this, many methods and software tools exist for comparative genomics using SNP- and gene-based phylogenetic analysis, either from unassembled sequencing reads, draft assemblies/contigs, or complete genome sequences. These methods are beyond the scope of this lesson.</p>
<div id="the-treeio-package" class="section level2">
<h2>The <code>treeio</code> Package</h2>
<p><strong>treeio</strong> is an R package that designed for phylogenetic tree data input and output. It is released as part of Bioconductor and ROpenSci projects.</p>
<ol style="list-style-type: decimal">
<li><strong>tree Bioconductor page</strong>: <a href="https://www.bioconductor.org/packages/treeio" class="uri">https://www.bioconductor.org/packages/treeio</a>.</li>
<li><strong>treeio homepage</strong>: <a href="https://guangchuangyu.github.io/treeio" class="uri">https://guangchuangyu.github.io/treeio</a>.</li>
</ol>
<p>Just like R packages from CRAN, you only need to install Bioconductor packages once (<a href="setup.html#bioconductor">instructions here</a>), then load them every time you start a new R session.</p>
<pre class="r"><code>library(treeio)</code></pre>
<hr />
</div>
<div id="getting-tree-data-from-evolutionary-analysis-result" class="section level2">
<h2>Getting tree data from evolutionary analysis result</h2>
<p>Most tree viewer software (including R packages) focus on <strong>Newick</strong> and <strong>NEXUS</strong> file formats, and other evolutionary analysis software might also contains supporting evidence and/or analysis findings within the file that can be further analyzed in R or interpreted in phylogenetic context to help identifying evolutionary patterns.</p>
<p><strong>treeio</strong> supports several file formats, including:</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Newick_format">Newick</a></li>
<li><a href="https://en.wikipedia.org/wiki/Nexus_file">Nexus</a></li>
<li><a href="https://home.cc.umanitoba.ca/~psgendb/doc/atv/NHX.pdf">New Hampshire eXtended format (NHX)</a></li>
<li><a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0031009">jplace</a></li>
<li><a href="https://en.wikipedia.org/wiki/PHYLIP">Phylip</a></li>
</ul>
<p>and software output from:</p>
<ul>
<li><a href="http://beast2.org/">BEAST</a></li>
<li><a href="http://sco.h-its.org/exelixis/web/software/epa/index.html">EPA</a></li>
<li><a href="https://veg.github.io/hyphy-site/">HyPhy</a></li>
<li><a href="http://nbisweden.github.io/MrBayes/">MrBayes</a></li>
<li><a href="http://abacus.gene.ucl.ac.uk/software/paml.html">PAML</a></li>
<li><a href="https://pbil.univ-lyon1.fr/software/phyldog/">PHYLDOG</a></li>
<li><a href="http://matsen.fhcrc.org/pplacer/">pplacer</a></li>
<li><a href="http://ginger.ucdavis.edu/r8s">r8s</a></li>
<li><a href="http://evomics.org/learning/phylogenetics/raxml/">RAxML</a></li>
<li><a href="https://revbayes.github.io/intro.html">RevBayes</a></li>
</ul>
<p>The <code>treeio</code> package implement several parser functions.</p>
<table>
<caption>Parser functions defined in treeio</caption>
<thead>
<tr class="header">
<th align="left">Parser function</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">read.beast</td>
<td align="left">parsing output of BEAST</td>
</tr>
<tr class="even">
<td align="left">read.codeml</td>
<td align="left">parsing output of CodeML (rst and mlc files)</td>
</tr>
<tr class="odd">
<td align="left">read.codeml_mlc</td>
<td align="left">parsing mlc file (output of CodeML)</td>
</tr>
<tr class="even">
<td align="left">read.hyphy</td>
<td align="left">parsing output of HYPHY</td>
</tr>
<tr class="odd">
<td align="left">read.jplace</td>
<td align="left">parsing jplace file including output of EPA and pplacer</td>
</tr>
<tr class="even">
<td align="left">read.mrbayes</td>
<td align="left">parsing output of MrBayes</td>
</tr>
<tr class="odd">
<td align="left">read.newick</td>
<td align="left">parsing newick string, with ability to parse node label as support values</td>
</tr>
<tr class="even">
<td align="left">read.nhx</td>
<td align="left">parsing NHX file including output of PHYLDOG and RevBayes</td>
</tr>
<tr class="odd">
<td align="left">read.paml_rst</td>
<td align="left">parsing rst file (output of BaseML or CodeML)</td>
</tr>
<tr class="even">
<td align="left">read.phylip</td>
<td align="left">parsing phylip file (phylip alignment + newick string)</td>
</tr>
<tr class="odd">
<td align="left">read.r8s</td>
<td align="left">parsing output of r8s</td>
</tr>
<tr class="even">
<td align="left">read.raxml</td>
<td align="left">parsing output of RAxML</td>
</tr>
</tbody>
</table>
<p>After parsing, storage of the tree structure with associated data is made through a S4 class, treedata, defined in the <a href="https://bioconductor.org/packages/treeio/">treeio</a> package. These parsed data are mapped to the tree branches and nodes inside <code>treedata</code> object, so that they can be efficiently used to visually annotate the tree using <a href="https://bioconductor.org/packages/ggtree/">ggtree</a> package.</p>
</div>

<div class="footer">
<a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank"><i class="fa fa-creative-commons fa-2x"></i></a>&nbsp;
<a href="https://twitter.com/guangchuangyu" target="_blank"><i class="fa fa-twitter fa-2x"></i></a>&nbsp;
<a href="https://github.com/GuangchuangYu/treedata-workshop" target="_blank"><i class="fa fa-github fa-2x"></i></a>&nbsp;
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
